templates/federated_reporting/parallel.sh

```diff
@@ -0,0 +1,127 @@
+# A shell "library" providing a function to run a given command on multiple
+# arguments in parallel.
+#
+# For best results, make sure 'parallel' is installed.
+#
+
+if type parallel >/dev/null; then
+  HAVE_PARALLEL=1
+else
+  HAVE_PARALLEL=0
+fi
+
+if type xargs >/dev/null; then
+  HAVE_XARGS=1
+else
+  HAVE_XARGS=0
+fi
+
+if type getconf >/dev/null; then
+  NCPUs="$(getconf _NPROCESSORS_ONLN)"
+else
+  NCPUs="$(grep processor /proc/cpuinfo | wc -l)"
+fi
+
+
+_run_using_parallel() {
+  max_jobs=""
+  if [ $# -gt 2 ]; then
+    max_jobs="-j$3"
+  fi
+
+  if [ "$2" = "-" ]; then
+    parallel $max_jobs "$1 {}"
+  else
+    parallel $max_jobs "$1 {}" :::: "$2"
+  fi
+  return $?
+}
+
+_run_using_xargs() {
+  if [ $# -gt 2 ]; then
+    max_jobs="-P$3"
+  else
+    max_jobs="-P$NCPUs"
+  fi
+
+  if [ "$2" = "-" ]; then
+    xargs -n1 $max_jobs "$1"
+  else
+    xargs -n1 $max_jobs -a "$2" "$1"
+  fi
+  return $?
+}
+
+_run_using_for() {
+  if [ $# -gt 2 ]; then
+    job_slots="$3"
+  else
+    job_slots="$NCPUs"
+  fi
+
+  if [ "$2" = "-" ]; then
+    input_arg="/dev/stdin"
+  else
+    input_arg="$2"
+  fi
+  failure=0
+  while read item; do
+    if [ "$job_slots" = 0 ]; then
+      wait -n
+      if [ $? != 0 ] && [ $failure = 0 ]; then
+        failure=1
+      fi
+      job_slots="$(expr $job_slots + 1)"
+    fi
+    "$1" "$item" &
+    job_slots="$(expr $job_slots - 1)"
+  done < $input_arg
+  wait
+  return $failure
+}
+
+run_in_parallel() {
```

Comment:
SHoudl this be called try_run_in_parallel since it falls back to a serial for loop in the end? really nitpicky, hard to find things to complain about.

Reply:
the for loop is using backgrounding `&` so it's still parallell. :)

Reply:
OK fine